### PR TITLE
fix: Do not crash when running sync client in async context

### PR DESCRIPTION
**Pull request by asgeirrr**
The asyncio code is now executed in a ThreadPoolExecutor to avoid problems with multiple event loops.

### CODE REVIEW

File: rossum_api/elis_api_client_sync.py [hunk 1]
```diff
@@ -75,29 +76,24 @@ def __init__(
...
-        ait = ait.__aiter__()
-        while True:
-            try:
-                obj = self.event_loop.run_until_complete(ait.__anext__())
-                yield obj
-            except StopAsyncIteration:
-                break
+        async def async_iter_to_list(ait: AsyncIterator[T]):
+            # TODO this materializes the whole generator into memory :(
+            return [obj async for obj in ait]
```
Comment (asgeirrr) [inline rossum_api/elis_api_client_sync.py:84]: This makes me sad but I didn't figure out how to use the previous pattern with ThreadPoolExectutor yet.

Comment (asgeirrr) [reply rossum_api/elis_api_client_sync.py]: I've managed to implement this via `queue.Queue` to communicate between asyncio in the thread and the sync code. Seems to work fine including error propagation (test for that added).

File: rossum_api/elis_api_client_sync.py [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 import asyncio
 import typing
+from concurrent.futures import ThreadPoolExecutor
+from queue import Queue as ThreadSafeQueue
```
Comment (valtervojta) [inline rossum_api/elis_api_client_sync.py:6]: Is there a reason for the rename?

File: rossum_api/elis_api_client_sync.py [hunk 3]
```diff
@@ -75,29 +77,40 @@ def __init__(
...
+        # Consume the queue until completion to retain the iterator nature even in sync context
+        while True:
+            item = queue.get()
+            if item is None:  # None is used to signal completion
                 break
+            yield item
+
+        future.result()
+
+    def _run_coroutine(self, couroutine):
```
Comment (valtervojta) [inline rossum_api/elis_api_client_sync.py:106]:
```suggestion
    def _run_coroutine(self, coroutine):
```

File: rossum_api/elis_api_client_sync.py [hunk 4]
```diff
@@ -75,29 +77,40 @@ def __init__(
...
+        future = self.executor.submit(asyncio.run, couroutine)
```
Comment (valtervojta) [inline rossum_api/elis_api_client_sync.py:107]:
```suggestion
        future = self.executor.submit(asyncio.run, coroutine)
```

File: rossum_api/elis_api_client_sync.py [hunk 5]
```diff
@@ -75,29 +77,40 @@ def __init__(
...
+        future = self.executor.submit(asyncio.run, async_iter_to_list(ait, queue))  # type: ignore
+
+        # Consume the queue until completion to retain the iterator nature even in sync context
+        while True:
+            item = queue.get()
+            if item is None:  # None is used to signal completion
                 break
+            yield item
+
+        future.result()
```
Comment (valtervojta) [inline rossum_api/elis_api_client_sync.py:104]: This is probably just an idea for future improvement, but should be rossum-api be able to set a total timeout?
We would have to replicate it in the sync client.

File: rossum_api/elis_api_client_sync.py [hunk 6]
```diff
@@ -75,29 +77,40 @@ def __init__(
...
-            except StopAsyncIteration:
+                async for obj in ait:
+                    queue.put(obj)
+            finally:
+                queue.put(None)  # Signal iterator was consumed
+
+        future = self.executor.submit(asyncio.run, async_iter_to_list(ait, queue))  # type: ignore
+
+        # Consume the queue until completion to retain the iterator nature even in sync context
+        while True:
```
Comment (valtervojta) [inline rossum_api/elis_api_client_sync.py:98]: Should we also have try finally block here to cancel the rest of the future in case sth goes wrong?

**Review by valtervojta — CHANGES_REQUESTED** [review]
I like it overall, pretty clean usage I think.

File: rossum_api/elis_api_client_sync.py (hunk 2 above)
Comment (asgeirrr) [reply rossum_api/elis_api_client_sync.py:6]: It clashes with our `Queue` dataclass.

File: rossum_api/elis_api_client_sync.py (hunk 6 above)
Comment (asgeirrr) [reply rossum_api/elis_api_client_sync.py:98]: I'm not sure if I understand. `async_iter_to_list` always puts `None` to the queue so I hope it should always correctly finish.

File: rossum_api/elis_api_client_sync.py (hunk 5 above)
Comment (asgeirrr) [reply rossum_api/elis_api_client_sync.py:104]: There are timeouts for individual HTTP requests which is probably enough.

File: rossum_api/elis_api_client_sync.py (hunk 6 above)
Comment (valtervojta) [reply rossum_api/elis_api_client_sync.py:99]: I suppose there is no other exception, no one is shutting down this queue.

**Review by valtervojta — APPROVED** [review]
LGTM

**Pull request merged**